File: typescript/src/decoder.ts
import { floatRead, utf8Read, RleReader } from "./serde";

export class Decoder {
  protected static _instance: Decoder | null = null;

  protected bytes!: Uint8Array;
  protected pos = 0;
  protected dict: string[] = [];
  protected rle = new RleReader();

  static create(buf: Uint8Array): Decoder {
    const dec = (Decoder._instance ??= new Decoder());
    dec.bytes = buf;
    dec.pos = 0;
    dec.dict = [];
    dec.rle.reset(buf);
    return dec;
  }

  protected constructor() {}

  nextString() {
    const lenOrIdx = this.readVarint();
    if (lenOrIdx === 0) {
      return "";
    }
    if (lenOrIdx > 0) {
      const str = this.readStringUtf8(lenOrIdx);
      this.dict.push(str);
      return str;
    }
    return this.dict[-lenOrIdx - 1]!;
  }

  nextInt() {
    return this.readVarint();
  }

  nextBoundedInt(min: number) {
    return this.readUVarint() + min;
  }

  nextFloat() {
    return this.readFloat();
  }

  nextFloatQuantized(precision: number) {
    return this.nextInt() * precision;
  }

  nextBoolean() {
    return this.rle.nextBit();
  }

  nextEnum(numBits: number): number {
    return this.rle.nextBits(numBits);
  }

  nextOptional<T>(innerRead: () => T): T | undefined {
    return this.nextBoolean() ? innerRead() : undefined;
  }

  nextArray<T>(innerRead: () => T): T[] {
    const len = this.readUVarint();
    const arr = new Array<T>(len);
    for (let i = 0; i < len; i++) {
      arr[i] = innerRead();
    }
    return arr;
  }

  nextRecord<K, T>(innerKeyRead: () => K, innerValRead: () => T): Map<K, T> {
    const len = this.readUVarint();
    const obj: Map<K, T> = new Map();
    for (let i = 0; i < len; i++) {
      obj.set(innerKeyRead(), innerValRead());
    }
    return obj;
  }

  protected readVarint() {
    const val = this.readUVarint();
    return val % 2 === 0 ? val / 2 : -(val + 1) / 2;
  }

  protected readUVarint() {
    let result = 0;
    let multiplier = 1;
    while (true) {
      const byte = this.bytes[this.pos++]!;
      result += (byte & 0x7f) * multiplier;
      if (byte < 0x80) {
        return result;
      }
      multiplier *= 128;
    }
  }

  protected readFloat() {
    const val = floatRead(this.bytes, this.pos);
    this.pos += 4;
    return val;
  }

  protected readStringUtf8(len: number): string {
    const start = this.pos;
    this.pos += len;
    return utf8Read(this.bytes, start, len);
  }
}

export class DiffDecoder extends Decoder {
  protected static override _instance: DiffDecoder | null = null;

  static override create(buf: Uint8Array): DiffDecoder {
    const dec = (DiffDecoder._instance ??= new DiffDecoder());
    dec.bytes = buf;
    dec.pos = 0;
    dec.dict = [];
    dec.rle.reset(buf);
    return dec;
  }

  nextStringDiff(a: string) {
    if (!this.dict.includes(a)) {
      this.dict.push(a);
    }
    return this.nextString();
  }

  nextIntDiff(_a: number) {
    return this.nextInt();
  }

  nextBoundedIntDiff(_a: number, min: number) {
    return this.nextBoundedInt(min);
  }

  nextFloatDiff(_a: number) {
    return this.nextFloat();
  }

  nextFloatQuantizedDiff(_a: number, precision: number) {
    return this.nextFloatQuantized(precision);
  }

  // Boolean diff is special - the change bit IS the diff
  nextBooleanDiff(a: boolean) {
    const changed = this.nextBoolean();
    return changed ? !a : a;
  }

  nextEnumDiff(_a: number, numBits: number): number {
    return this.nextEnum(numBits);
  }

  // Object diff - reads change bit and returns old or decoded object
  nextObjectDiff<T>(a: T, decodeDiff: () => T): T {
    return this.nextBoolean() ? decodeDiff() : a;
  }

  // Generic field diff - reads change bit and returns old or new value
  nextFieldDiff<T>(a: T, decodeDiff: (a: T) => T): T {
    return this.nextBoolean() ? decodeDiff(a) : a;
  }

  nextOptionalDiff<T>(obj: T | undefined, decode: () => T, decodeDiff: (a: T) => T): T | undefined {
    // Optimization: if obj was null, we know new value must be non-null (else changed would be false)
    // So skip reading the present bit in null→value case
    if (obj == null) {
      return decode();
    } else {
      const present = this.nextBoolean();
      return present ? decodeDiff(obj) : undefined;
    }
  }

  nextArrayDiff<T>(arr: T[], decode: () => T, decodeDiff: (a: T) => T): T[] {
    const newLen = this.readUVarint();

    // Start with copy of old array (truncated to new length)
    const newArr = arr.slice(0, Math.min(arr.length, newLen));

    // Apply updates (sparse)
    const numUpdates = this.readUVarint();
    for (let i = 0; i < numUpdates; i++) {
      const idx = this.readUVarint();
      newArr[idx] = decodeDiff(arr[idx]!);
    }

    // Read additions
    for (let i = arr.length; i < newLen; i++) {
      newArr.push(decode());
    }

    return newArr;
  }

  nextRecordDiff<K, T>(obj: Map<K, T>, decodeKey: () => K, decodeVal: () => T, decodeDiff: (a: T) => T): Map<K, T> {
    const result: Map<K, T> = new Map(obj);
    if (obj.size > 0) {
      const numDeletions = this.readUVarint();
      for (let i = 0; i < numDeletions; i++) {
        const key = decodeKey();
        result.delete(key);
      }
      const numUpdates = this.readUVarint();
      for (let i = 0; i < numUpdates; i++) {
        const key = decodeKey();
        result.set(key, decodeDiff(result.get(key)!));
      }
    }
    const numAdditions = this.readUVarint();
    for (let i = 0; i < numAdditions; i++) {
      const key = decodeKey();
      const val = decodeVal();
      result.set(key, val);
    }
    return result;
  }
}

File: csharp/src/Decoder.cs
using System.Text;

namespace DeltaPack;

public class Decoder
{
    [ThreadStatic]
    private static RleReader? _sharedRle;

    [ThreadStatic]
    private static List<string>? _sharedDict;

    private readonly byte[] _buffer;
    private int _pos;
    private readonly List<string> _dict;
    private readonly RleReader _rle;

    /// <summary>
    /// Tracker that decoded tracked-container instances (and tracked-class results) bind to.
    /// Set by the entry-point Decode/DecodeDiff methods so that decoded values land in the
    /// correct domain.
    /// </summary>
    public Tracker Tracker { get; set; } = Tracker.Default;

    public Decoder(byte[] buf)
    {
        _buffer = buf;
        _pos = 0;
        _rle = _sharedRle ??= new RleReader();
        _rle.Reset(buf);
        _dict = _sharedDict ??= new List<string>();
        _dict.Clear();
    }

    // Primitive methods

    public string NextString()
    {
        var lenOrIdx = NextInt();
        if (lenOrIdx == 0)
            return "";
        if (lenOrIdx > 0)
        {
            var str = Encoding.UTF8.GetString(_buffer, _pos, (int)lenOrIdx);
            _pos += (int)lenOrIdx;
            _dict.Add(str);
            return str;
        }
        return _dict[(int)(-lenOrIdx - 1)];
    }

    public long NextInt() =>
        Varint.ReadVarint(_buffer, ref _pos);

    public long NextBoundedInt(long min) =>
        (long)NextUInt() + min;

    public ulong NextUInt() =>
        Varint.ReadUVarint(_buffer, ref _pos);

    public float NextFloat()
    {
        var val = BitConverter.ToSingle(_buffer, _pos);
        _pos += 4;
        return val;
    }

    public float NextFloatQuantized(float precision) =>
        NextInt() * precision;

    public bool NextBoolean() =>
        _rle.NextBit();

    public int NextEnum(int numBits) =>
        _rle.NextBits(numBits);

    // Container methods

    public T? NextOptional<T>(Func<T> innerRead) where T : class =>
        NextBoolean() ? innerRead() : null;

    public List<T> NextArray<T>(Func<T> innerRead)
    {
        var len = (int)NextUInt();
        var arr = new List<T>(len);
        for (var i = 0; i < len; i++)
            arr.Add(innerRead());
        return arr;
    }

    public DPDict<TKey, TValue> NextRecord<TKey, TValue>(
        Func<TKey> innerKeyRead,
        Func<TValue> innerValRead)
        where TKey : notnull
    {
        var len = (int)NextUInt();
        var dict = new DPDict<TKey, TValue>(Tracker, len);
        for (var i = 0; i < len; i++)
            dict[innerKeyRead()] = innerValRead();
        return dict;
    }

    // Diff methods

    // Value-only diff methods (caller handles change bit for object fields)

    public string NextStringDiff(string a)
    {
        if (!_dict.Contains(a))
            _dict.Add(a);
        return NextString();
    }

    public long NextIntDiff(long a) =>
        NextInt();

    public long NextBoundedIntDiff(long a, long min) =>
        NextBoundedInt(min);

    public float NextFloatDiff(float a) =>
        NextFloat();

    public float NextFloatQuantizedDiff(float a, float precision) =>
        NextFloatQuantized(precision);

    public bool NextBooleanDiff(bool a)
    {
        var changed = NextBoolean();
        return changed ? !a : a;
    }

    public int NextEnumDiff(int a, int numBits) =>
        NextEnum(numBits);

    // Object diff helper (read change bit, decode if changed)

    public T NextObjectDiff<T>(T a, Func<T> decodeDiff)
    {
        var changed = NextBoolean();
        return changed ? decodeDiff() : a;
    }

    // Field diff helper (read change bit, decode if changed)

    public T NextFieldDiff<T>(T a, Func<T, T> decodeDiff)
    {
        var changed = NextBoolean();
        return changed ? decodeDiff(a) : a;
    }

    // Optional diff

    public T? NextOptionalDiff<T>(T? a, Func<T> decode, Func<T, T> decodeDiff) where T : class
    {
        // Optimization: if a was null, we know b must be non-null (else changed would be false)
        // So no present bit in null→value case
        if (a is null)
        {
            return decode(); // null → value (guaranteed non-null by caller)
        }
        else
        {
            var present = NextBoolean();
            return present ? decodeDiff(a) : null;
        }
    }

    public T? NextOptionalDiff<T>(T? a, Func<T> decode, Func<T, T> decodeDiff) where T : struct
    {
        // Optimization: if a was null, we know b must be non-null (else changed would be false)
        // So no present bit in null→value case
        if (a is null)
        {
            return decode(); // null → value (guaranteed non-null by caller)
        }
        else
        {
            var present = NextBoolean();
            return present ? decodeDiff(a.Value) : null;
        }
    }

    public List<T> NextArrayDiff<T>(IList<T> arr, Func<T> decode, Func<T, T> decodeDiff)
    {
        // Caller handles change bit via NextFieldDiff
        var newLen = (int)NextUInt();

        // Start with copy of old array (truncated to new length)
        var newArr = arr.Take(Math.Min(arr.Count, newLen)).ToList();

        // Apply updates (sparse)
        var numUpdates = (int)NextUInt();
        for (var i = 0; i < numUpdates; i++)
        {
            var idx = (int)NextUInt();
            newArr[idx] = decodeDiff(arr[idx]);
        }

        // Read additions
        for (var i = arr.Count; i < newLen; i++)
            newArr.Add(decode());

        return newArr;
    }

    public DPDict<TKey, TValue> NextRecordDiff<TKey, TValue>(
        DPDict<TKey, TValue> obj,
        Func<TKey> decodeKey,
        Func<TValue> decodeVal,
        Func<TValue, TValue> decodeDiff)
        where TKey : notnull
    {
        // Build the result by copying the baseline directly (bypassing the user-facing
        // aliasing guard on the ctor). The decoder owns both obj and the returned result,
        // so moving tracked children from one to the other is legitimate.
        var result = DPDict<TKey, TValue>.CreateSnapshot(obj);

        if (obj.Count > 0)
        {
            var numDeletions = (int)NextUInt();
            for (var i = 0; i < numDeletions; i++)
            {
                var key = obj.GetKeyAtIndex((int)NextUInt());
                result.Remove(key);
            }

            var numUpdates = (int)NextUInt();
            for (var i = 0; i < numUpdates; i++)
            {
                var key = obj.GetKeyAtIndex((int)NextUInt());
                result[key] = decodeDiff(result[key]);
            }
        }

        var numAdditions = (int)NextUInt();
        for (var i = 0; i < numAdditions; i++)
        {
            var key = decodeKey();
            var val = decodeVal();
            result[key] = val;
        }

        return result;
    }

    public DPList<T> NextArrayDiff<T>(DPList<T> arr, Func<T> decode, Func<T, T> decodeDiff)
    {
        var newLen = (int)NextUInt();
        var copyLen = Math.Min(arr.Count, newLen);
        // Prefill via the internal copy (same reasoning as NextRecordDiff) so the aliasing
        // guard on Add() doesn't fire on children still parented to `arr`.
        var result = DPList<T>.CopyPrefixForDecode(arr, copyLen, newLen, Tracker);

        var numUpdates = (int)NextUInt();
        for (var i = 0; i < numUpdates; i++)
        {
            var idx = (int)NextUInt();
            result[idx] = decodeDiff(arr[idx]);
        }

        for (var i = arr.Count; i < newLen; i++)
            result.Add(decode());

        return result;
    }
}
